### PR TITLE
push class file names into ClasspathEntry

### PR DESCRIPTION
**Pull request by elharo**
some initial work on pushing more info into classes instead of parallel data structures

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java [hunk 1]
```diff
@@ -48,22 +48,18 @@
 
   /** An entry for a Maven artifact. */
   public ClassPathEntry(Artifact artifact) {
-    checkNotNull(artifact.getFile());
+    this(artifact.getFile().toPath());
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java]: This defies line 39 "Either jar or artifact is non-null".

Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java]: removed

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 2]
```diff
@@ -189,7 +189,10 @@ private static Path generate(Bom bom) throws IOException, TemplateException, URI
     // When checking a BOM, entry point classes are the ones in the artifacts listed in the BOM
     List<ClassPathEntry> artifactJarsInBom = classpath.subList(0, managedDependencies.size());
     ImmutableSet<ClassPathEntry> entryPoints = ImmutableSet.copyOf(artifactJarsInBom);
-
+    for (ClassPathEntry entry : entryPoints) {
+      entry.loadClassFileNames();
+    }
+
```
Comment (suztomo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:195]: I feel this piece of code is wrong.  It is as if `classPathResult.getClassPath()` has a contract that its caller needs to call `loadClassFileNames()` for each elements.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java [hunk 3]
```diff
@@ -96,27 +103,31 @@ public String toString() {
...
 
-    return classPath.getAllClasses().stream().map(ClassInfo::getName).collect(toImmutableSet());
+    classFileNames =
+        classPath.getAllClasses().stream().map(ClassInfo::getName).collect(toImmutableSet());
+  }
+  
+  /**
+   * @return the fully qualified names of the classes in this entry's jar file
+   */
+  public ImmutableSet<String> getClassNames() {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java:130]: How about calling `loadClassFileNames` and cache the result?

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 2 above)
Comment (elharo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java:195]: I don't like this either I haven't found something I like better yet though. I was hoping you might have an idea.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java (hunk 3 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java:130]: I tried this. The problem is it introduces an IOException on this method, which both feels wrong and makes a lot of calling code very ugly. 

I'm not sure how to proceed. Suggestions appreciated.

Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java:130]: Maybe the right place to do this is in the constructor. It's called much less frequently than `getClassNames` so adding an IOException there would be less invasive.

**Comment by suztomo** [comment]
For making the JAR-reading at instance creation,
- factory methods are better choice as in  "Avoid doing work in constructors" go/java-practices/constructors#avoid-doing-work-in-constructors . (Same 
- It makes tests harder. I asssume `ClassPathEntry.of("foo:bar:1.2.3", "dummy.jar")` starts throwing IOException.


I'm feeling getClassNames should do the IO work.

>  The problem is it introduces an IOException on this method, which both feels wrong and makes a lot of calling code very ugly.

As the original method `mapJarToClassFileNames` (being removed) has IOException, it shouldn't create new problems with IOExceptions. I'm checking how it looks like.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 4]
```diff
@@ -73,7 +69,6 @@
   private final ImmutableList<ClassPathEntry> inputClassPath;
   private final FixedSizeClassPathRepository classRepository;
   private final ClassLoader extensionClassLoader;
-  private final ImmutableSetMultimap<ClassPathEntry, String> classPathEntryToClassFileNames;
   private final ImmutableListMultimap<String, ClassPathEntry> classFileNameToClassPathEntry;
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:72]: As much as possible needs to be pushed into ClasspathEntry, and we need to start replacing all these parallel multimaps with object types. The existing structure is brittle and hard to follow.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 5]
```diff
@@ -92,19 +87,25 @@ static ClassDumper create(List<ClassPathEntry> entries) throws IOException {
             .collect(toImmutableList());
     checkArgument(
         unreadableFiles.isEmpty(), "Some jar files are not readable: %s", unreadableFiles);
-
-    return new ClassDumper(entries, extensionClassLoader, mapJarToClassFileNames(entries));
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:96]: Let me think about that. There might be another way to do this, but it can't involve maintaining multiple multimaps.

I do like the principle of keeping work out of constructors. However I don't believe one should contort the natural design of a class to support testing and mocking. That leads to both bad design and tests that don't actually test the real code. Most of the time more effort spent in test setup avoid the need for mocking.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java [hunk 6]
```diff
@@ -118,4 +121,11 @@ public static ClassPathEntry of(String coordinates, String filePath) {
 
     return classPath.getAllClasses().stream().map(ClassInfo::getName).collect(toImmutableSet());
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java:122]: Memo: this `classPath.getAllClasses()` returns empty if `jar` does not exist.
Screen Shot 2020-04-02 at 11 42 39

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassPathEntry.java:122]: If we start reading this JAR files without Guava's ClassPath (https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1320 ). I think appropriate behavior is to throw IOException upon nonexistent JAR file.

The existing tests didn't fail for this PR because Guava's ClassPath happened to ignore nonexistent JAR file.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 7]
```diff
@@ -73,7 +69,6 @@
   private final ImmutableList<ClassPathEntry> inputClassPath;
   private final FixedSizeClassPathRepository classRepository;
   private final ClassLoader extensionClassLoader;
-  private final ImmutableSetMultimap<ClassPathEntry, String> classPathEntryToClassFileNames;
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:76]: Memo: where does LinkageChecker use this ?

It's used by `classDumper.classNamesInJar()`. This method is used by Linkage Checker:

```
    classToFieldSymbols.forEach(
        (classFile, fieldSymbol) -> {
          if (!classDumper
              .classNamesInJar(classFile.getClassPathEntry())
              .contains(fieldSymbol.getClassBinaryName())) {
            findSymbolProblem(classFile, fieldSymbol)
                .ifPresent(problem -> problemToClass.put(problem, classFile.topLevelClassFile()));
          }
        });
```

This if-statement checks whether the source class and target class of a symbol reference are in the same JAR file or not. If they are in the same JAR file, Linkage Checker skips the references because such javac would not compile to generate such JAR files.

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java:76]: Therefore the result of class file name reading should be cached (this PR does that)

**Comment by suztomo** [comment]
For 1-3, they make sense. Although I liked symmetry of many-to-many relationship, I also see the benefit of pushing class file names to ClassPathEntry. Now the discussion between reading JAR in a constructor or a factory method.

For 4,

> We should not contort the design of a class to make testing easier.

We should aim to avoid making tests less testable. Reading JAR files in the ClassPathEntry constructor makes the class less testable (below):

> This PR does not make this class any less testable. All existing tests pass.

The existing tests didn't fail for this PR because Guava's ClassPath _happened to_ not raise an IOException upon nonexistent JAR files.

If we start reading this JAR files without Guava's ClassPath (#1320 ), then the constructor starts throwing IOException (I strongly believe this is correct behavior of the `readClassFileNames `). In tests with "dummy.jar", the class would require mockito or some special handling to avoid raising IOException. ["Avoid doing work in constructors
"](https://engdoc.corp.google.com/eng/doc/devguide/java/practices/constructors.md?cl=head#avoid-doing-work-in-constructors ) explains simple constructors let us write tests without mockito.

For 5,

the discussion in "Avoid doing work in constructors
"](https://engdoc.corp.google.com/eng/doc/devguide/java/practices/constructors.md?cl=head#avoid-doing-work-in-constructors ) does not require subclasses. It's fine to use static factory method without subclasses.

**Comment by suztomo** [comment]
Surveyed the design of other classes:

- No IO at constructor: `java.nio.file.Path`, `java.io.File`, `DefaultArtifact`
- Potential IOException at constructor: `java.util.jar.JarFile`.

I recognize ClassPathEntry is similar to the former group. The creation of the class does not mean reading the content. The former group may have methods (or helper class) to read their content but not at the constructor.

**Comment by elharo** [comment]
I don;t think you can compare java.io.File to this class. Java.io.File has always been a bit of a weird duck, going back to Java 1.0, and was so badly designed Oracle had to go back to the drawing board and do it again with Path years later.

I think we want ClasspathEntry to actually exist, not simply be a pointer to something that might or might not exist. 

Ultimately there will be a point at which we find out a file can't be read and need to handle the failure. That's unescapable. The question is when is it most convenient to recognize that? About the earliest we can do that is when the ClasspathEntry is constructed, and indeed we already fail fast at that point if the jar is missing, just with a NullPointerException instead of an IOException. The latest we can do that is when we're trying to list out the contents of the jar, which is essentially what we're doing now. 

Failing early is usually easier to fix but failing late might avoid some cases where we don't need to read the JAR at all.
